scripts/x-api.js

```diff
@@ -0,0 +1,577 @@
+            'https://w3id.org/xapi/video/extensions/time': resultExtTime
+          }
+        },
+        'context': {
+          'contextActivities': {
+            'category': [{
+              'id': 'https://w3id.org/xapi/video'
+            }]
+          },
+          'extensions': extensions
+        },
+        'timestamp' : timeStamp
+      };
+    };
+
+    /**
+     * Generates 'completed' xAPI statement (Video Profile).
+     * @see https://liveaspankaj.gitbooks.io/xapi-video-profile/content/statement_data_model.html#236-completed
+     *
+     * @public
+     * @param {Number} currentTime time of the video currently
+     * @param {Number} duration length of the current video in seconds
+     * @param {Number} progress Number between 0.000 and 1.000 indicating percentage of video watched
+     * @returns {Object} JSON xAPI statement
+     */
+    self.getArgsXAPICompleted = function (currentTime, duration, progress) {
+      var resultExtTime = formatFloat(currentTime);
+      var dateTime = new Date();
+      endPlayingSegment(resultExtTime);
+      playingSegmentStart = 0;
+      var timeStamp = dateTime.toISOString();
+
+      var extensions = {};
+      if (typeof resultExtTime !== 'undefined') {
+        extensions['https://w3id.org/xapi/video/extensions/time'] = resultExtTime;
+      }
+      if (typeof progress !== 'undefined') {
+        extensions['https://w3id.org/xapi/video/extensions/progress'] = progress;
+      }
+      if (typeof playedSegments !== 'undefined') {
+        extensions['https://w3id.org/xapi/video/extensions/played-segments'] = stringifyPlayedSegments();
+      }
+
+      return {
+        'verb': {
+          'id': 'http://adlnet.gov/expapi/verbs/completed',
+          'display': {
+            'en-US': 'completed'
+          }
+        },
+        'object': getXAPIObject(),
+        'result': {
+          'extensions': extensions,
+          'completion' : true,
+          'duration' : secondsToISO8601Duration(duration)
+        },
+        'context': {
+          'contextActivities': {
+            'category': [{
+              'id': 'https://w3id.org/xapi/video'
+            }]
+          },
+          'extensions': {
+            'https://w3id.org/xapi/video/extensions/session-id': sessionID
+          }
+        },
+        'timestamp' : timeStamp
+      };
+    };
+
+    /**
+     * Calculate video progress.
+     *
+     * @public
+     * @param {Number} currentTime current time of the video in seconds
+     * @param {Number} duration length of the video in seconds
+     * @returns {Number} Progress between 0..1
+     */
+    self.getProgress = function (currentTime, duration) {
+      // If we're currently playing a segment, end it so it's included in our
+      // calculations below.
+      endPlayingSegment(currentTime);
+
+      // Create a copy of the played segments array so we can manipulate it.
+      var parsedPlayedSegments = JSON.parse(JSON.stringify(playedSegments));
```

Comment:
I think this looks much cleaner, thanks!

I will change this later myself if you don't have the time, but if you cannot work with a reference to an array, you can also clone slice() which might be easier and quicker than going for JSON parsing.

`var parsedPlayedSegments = playedSegments.slice();`

Reply:
I actually started with `playedSegments.slice()`, but since our array members are objects (`{start:0, end:0}`), `slice()` will only clone the array, and still use references to the original objects (shallow clone).
See:
https://stackoverflow.com/questions/42523881/how-to-clone-a-javascript-array-of-objects/42523934
https://stackoverflow.com/questions/7486085/copying-array-by-value-in-javascript#answer-23536726

Reply:
Right, that was a deep array ... Sorry.

Reply:
No problem! I've got some time set aside today to go over the rest of the review comments, will keep you posted. (Likely we'll address some of them, and punt the more complicated ones to you folks...but we'll see! :)